modules/reactor/src/meshgenerators/ExtraElementIDConnectionFixup.C

```diff
@@ -0,0 +1,231 @@
+          break;
+        }
+
+      continue;
+    }
+
+    // fine element ids that has been visited
+    std::set<dof_id_type> visited;
+    // fine element buffer for visiting
+    std::vector<const Elem *> buffer;
+    // nodes of fine elements that have been visited
+    std::set<dof_id_type> visited_nodes;
+    // a copy of fine elements to track unvisited elements
+    auto fine_ids_copy = fine_ids;
+
+    while (!fine_ids_copy.empty())
+    {
+      buffer.push_back(mesh->elem_ptr(*fine_ids_copy.begin()));
+
+      // nodes of all elements in this current visit
+      std::set<dof_id_type> nodes;
+      while (buffer.size() != 0)
+      {
+        const Elem * elem = buffer.back();
+        visited.insert(elem->id());
+        fine_ids_copy.erase(elem->id());
+        buffer.pop_back();
+
+        for (unsigned int n = 0; n < elem->n_nodes(); ++n)
+          nodes.insert(elem->node_ptr(n)->id());
+
+        // push connected elements through sides with the same coarse ID into the buffer
+        for (auto neig : elem->neighbor_ptr_range())
+          if (neig && neig->get_extra_integer(src_id) == coarse_id &&
+              visited.count(neig->id()) == 0)
+            buffer.push_back(neig);
+      }
+
+      // merge nodes
+      if (!visited_nodes.empty())
+      {
+        std::set<dof_id_type> int_nodes;
+        std::set_intersection(visited_nodes.begin(),
+                              visited_nodes.end(),
+                              nodes.begin(),
+                              nodes.end(),
+                              std::inserter(int_nodes, int_nodes.begin()));
+
+        if (int_nodes.size() == 0)
+          mooseError("Coarse elements with ID ", coarse_id, " has disconnected parts!");
+
+        // we will do a simple merge by assigning all elements connecting to the intersection nodes
+        // with the current coarse element ID
+        // FIXME: checking the connection through line segments in 3D can avoid some unnecessary
+        // assignment.
+        for (const auto & id : int_nodes)
+          for (const auto & elem_id : map_node_to_fine[id])
+          {
+            Elem * elem = mesh->elem_ptr(elem_id);
+            auto old_id = elem->get_extra_integer(src_id);
+            if (old_id != coarse_id)
+            {
+              if (_verbose)
+                _console << " Element with ID " << elem->id() << ": " << old_id << " => "
+                         << coarse_id << std::endl;
+              elem->set_extra_integer(src_id, coarse_id);
+              map_coarse_to_fine[old_id].erase(elem->id());
+              map_coarse_to_fine[coarse_id].insert(elem->id());
+              for (unsigned int n = 0; n < elem->n_nodes(); ++n)
+                visited_nodes.insert(elem->node_ptr(n)->id());
+
+              // put changed ids into re-consideration
+              coarse_ids.insert(old_id);
+            }
+          }
+      }
+      for (auto & id : nodes)
+        visited_nodes.insert(id);
+    }
+  }
```

Comment:
this whole section feels convoluted. Maybe it is optimal code that does what is desired exactly. 
With more comments I should be able to tell

Reply:
I think the major piece of comments missing is the above call `set_intersection`. I'v also added more.